webapp/src/snippetBuilder.tsx

```diff
@@ -0,0 +1,339 @@
+            answers: {},
+            currentQuestion: 0, // Index to track current question
+            defaults: {},
+            config: staticConfig,
+            output: staticConfig.output
+        };
+
+        this.hide = this.hide.bind(this);
+        this.cancel = this.cancel.bind(this);
+        this.confirm = this.confirm.bind(this);
+        this.backPage = this.backPage.bind(this);
+        this.nextPage = this.nextPage.bind(this);
+    }
+
+    buildDefaults() {
+        const { config } = this.state;
+        const defaults: any = {};
+
+        for (const question of config.questions) {
+            const { inputs } = question;
+            for (const input of inputs) {
+                const { defaultAnswer, answerToken } = input;
+                defaults[answerToken] = defaultAnswer;
+            }
+        }
+
+        this.setState({ defaults });
+    }
+
+    componentDidMount() {
+        // Sets default values
+        this.buildDefaults();
+    }
+
+    replaceTokens(output: string) {
+        const { answers, defaults } = this.state;
+        let tokenizedOutput = output;
+        const tokens = Object.keys(defaults);
+
+        // Replaces output tokens with answer if available or default value
+        for (let token of tokens) {
+            if (answers[token]) {
+                tokenizedOutput = tokenizedOutput.split(`$${token}`).join(answers[token]);
+            }
+            else {
+                tokenizedOutput = tokenizedOutput.split(`$${token}`).join(defaults[token]);
+            }
+        }
+
+        return tokenizedOutput;
+    }
+
+    generateOutputMarkdown(output: string) {
+        const { config } = this.state;
+        // Attaches starting and ending line based on output type
+        let md = `\`\`\`${config.outputType}\n`;
+        md += this.replaceTokens(output);
+        md += `\n\`\`\``;
+
+        return md
+    }
+
+    hide() {
+        this.setState({
+            visible: false
+        });
+    }
+
+    show(projectView: pxt.editor.IProjectView, mainWorkspace: Blockly.Workspace) {
+        pxt.tickEvent('snippetBuilder.show', null, { interactiveConsent: false });
+        this.setState({
+            visible: true,
+            mainWorkspace,
+            projectView,
+        });
+    }
+
+    cancel() {
+        pxt.tickEvent("snippetBuilder.cancel", undefined, { interactiveConsent: true });
+        this.hide();
+    }
+
+    injectBlocksToWorkspace() {
+        const { mainWorkspace, output } = this.state;
+
+        compiler.getBlocksAsync()
+            .then(blocksInfo => compiler.decompileBlocksSnippetAsync(this.replaceTokens(output), blocksInfo))
+            .then(resp => {
+                Blockly.Xml.appendDomToWorkspace(Blockly.Xml.textToDom(resp), mainWorkspace);
+            });
+    }
+
+    confirm() {
+        this.injectBlocksToWorkspace();
+        this.hide();
+    }
+
+    changePage(increment: 1 | -1) {
+        const { currentQuestion } = this.state;
```

Comment:
might be worth adding a tick here because we often see >80% of people not getting past first page for things like this